File: src/go/two_pointer/42.trapping-rain-water.go
package main

/*
 * @lc app=leetcode id=42 lang=golang
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
func trap(height []int) int {
	n := len(height)
	if n < 3 {
		return 0
	}
	l := 0
	r := n - 1
	leftMax := height[l]
	rightMax := height[r]
	trapped := 0
	for l < r {
		if leftMax < rightMax {
			l++
			leftMax = max(leftMax, height[l])
			trapped += leftMax - height[l]
		} else {
			r--
			rightMax = max(rightMax, height[r])
			trapped += rightMax - height[r]
		}
	}
	return trapped
}

// @lc code=end

File: src/java/two_pointer/42.trapping-rain-water.java
class P42 {
    // @lc app=leetcode id=42 lang=java
    // @lc code=start
    public class Solution {
        public int trap(int[] height) {
            int n = height.length;
            int i = 0, j = n - 1, leftMax = 0, rightMax = 0, water = 0;
            while (i <= j) {
                if (leftMax <= rightMax) {
                    leftMax = Math.max(leftMax, height[i]);
                    water += leftMax - height[i];
                    i++;
                } else {
                    rightMax = Math.max(rightMax, height[j]);
                    water += rightMax - height[j];
                    j--;
                }
            }
            return water;
        }
    }
    // @lc code=end
}
